1-Draft/RFCNNNN-Make-Terminating-Errors-Terminate-The-Right-Way.md

```diff
@@ -0,0 +1,169 @@
+
+```PowerShell
+& {
+    1/0
+    'Why?'
+}
+```
+
+PowerShell has upheld this behaviour since version 1.0 of the language. You can make the terminating error actually terminate execution of the command, by wrapping the command in try/catch, like this:
+
+```PowerShell
+try {
+    1/0
+    'Why?'
+} catch {
+    throw
+}
+```
+
+In that example, the exception raised by dividing by zero properly terminates execution of the running command.
+
+The difference between these two examples poses a risk to scripters who share scripts or modules with the community. The risk is that end users using a shared resource such as a script or module may see different behaviour from the logic within that module depending on whether or not they were inside of a `try` block when they invoked the script or a command exported by the module. That risk is very undesirable, and as a result many community members who share scripts/modules with the community wrap their logic in a `try/catch{throw}` (or similar) scaffolding to ensure that the behavior of their code is consistent no matter where or how it was invoked.
+
+Now consider this code snippet:
+
+```PowerShell
+New-Module -Name ThisShouldNotImport {
+    $myList = [System.Collections.Generics.List[string]]::new()
+
+    function Test-RFC {
+        [CmdletBinding()]
+        param()
+        'Some logic'
+        1 / 0 # Oops
+        'Some more logic'
+    }
+} | Import-Module
+```
+
+If you invoke that snippet, the `ThisShouldNotImport` module imports successfully because the terminating error (`[System.Collections.Generics.List[string]]` is not a valid type name) does not actually terminate the loading of the module. This could cause your module to load in an unexpected state, which is a bad idea. If you loaded your module by invoking a command defined with that module, you won't see the terminating error that was raised during the loading of the module (the terminating error that was raised during the loading of the module is not shown at all in that scenario!), so you could end up with some undesirable behaviour when that command executes even though the loading of the module generated a "terminating" error, and not have a clue why. Further, the Test-RFC command exported by this module produces a terminating error, yet continues to execute after that error. Last, if the caller either loads your module or invokes your command inside of a `try` block, they will see different behaviour. Any execution of code beyond a terminating error should be intentional, not accidental like it is in both of these cases, and it most certainly should not be influenced by whether or not the caller loaded the module or invoked the command inside of a `try` block. Binary modules do not behave this way. Why should script modules be any different?
```

Comment:
I'd recommend adding single line breaks after the end of every sentence (in the whole document). It'll make reading & reviewing the RFCs much easier (letting folks comment/respond to each sentence where needed), and won't affect the actual MD doc rendering (single line breaks are ignored).

Also, super nit but this is a bit of a lengthy paragraph. Maybe a paragraph break before `Further, if the Test-RFC ...` would make this a bit easier to read as well.

Reply:
Ugh. I really dislike line breaks for each sentence. Even though Markdown supports it, and keeps it a paragraph, writing that way is...ugh....hard for me to do. That doesn't mean I won't get there...it just means I'm not there yet.

That said, this paragraph was obviously a brain dump and should have been a list of points, so it has been converted accordingly. Thanks. :)